File: chapter-4-crypto-cryptography/sample/randomness/mersenne-twister.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	// Using time as seed (NOT secure for cryptography!)
	rand.Seed(time.Now().UnixNano())

	fmt.Println("Mersenne Twister Random Number Generation")
	fmt.Println("==========================================\n")

	// Generate some random numbers
	fmt.Println("Random integers (0-100):")
	for i := 0; i < 5; i++ {
		fmt.Printf("  %d: %d\n", i+1, rand.Intn(100))
	}

	fmt.Println("\nRandom floats (0.0-1.0):")
	for i := 0; i < 5; i++ {
		fmt.Printf("  %d: %.6f\n", i+1, rand.Float64())
	}

	// Demonstrate reproducibility with same seed
	fmt.Println("\nDemonstrating reproducibility with seed=42:")
	rand.Seed(42)
	fmt.Println("First run:")
	for i := 0; i < 3; i++ {
		fmt.Printf("  %d\n", rand.Intn(1000))
	}

	rand.Seed(42)
	fmt.Println("Second run (same seed):")
	for i := 0; i < 3; i++ {
		fmt.Printf("  %d\n", rand.Intn(1000))
	}

	fmt.Println("\n⚠️  WARNING: This is NOT cryptographically secure!")
	fmt.Println("For cryptographic use, use crypto/rand instead:")
	fmt.Println("  import \"crypto/rand\"")
	fmt.Println("  bytes := make([]byte, 32)")
	fmt.Println("  crypto/rand.Read(bytes)")
}

File: chapter-3-crypto-cryptography/sample/randomness/mersenne-twister.py
#!/usr/bin/env python3
"""
Mersenne Twister Random Number Generation Example
Python's random module uses the Mersenne Twister algorithm by default.
"""

import random

def main():
    print("Mersenne Twister Random Number Generation")
    print("==========================================\n")

    # Generate some random numbers with current time as seed
    print("Random integers (1-100):")
    for i in range(5):
        print(f"  {i+1}: {random.randint(1, 100)}")

    print("\nRandom floats (0.0-1.0):")
    for i in range(5):
        print(f"  {i+1}: {random.random():.6f}")

    # Demonstrate reproducibility with same seed
    print("\nDemonstrating reproducibility with seed=42:")
    random.seed(42)
    print("First run:")
    for i in range(3):
        print(f"  {random.randint(1, 1000)}")

    random.seed(42)
    print("Second run (same seed):")
    for i in range(3):
        print(f"  {random.randint(1, 1000)}")

    # Show different random methods
    print("\nOther random functions:")
    random.seed()  # Reset seed
    colors = ['red', 'blue', 'green', 'yellow', 'purple']
    print(f"  Random choice from list: {random.choice(colors)}")
    print(f"  Random sample of 3: {random.sample(colors, 3)}")
    print(f"  Gaussian distribution (μ=0, σ=1): {random.gauss(0, 1):.4f}")

    print("\n⚠️  WARNING: This is NOT cryptographically secure!")
    print("For cryptographic use, use the secrets module instead:")
    print("  import secrets")
    print("  token = secrets.token_hex(32)")
    print("  secure_number = secrets.randbelow(100)")

if __name__ == "__main__":
    main()
